pkg/tfbridge/walk.go

```diff
@@ -168,3 +284,394 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
+	// If the visitor has done something.
+	//
+	// Visits will only be replayed during runtime startup if `HasEffect: true`.
+	HasEffect bool
+}
+
+func (v *VisitInfo) SchemaPath() SchemaPath                         { return v.schemaPath }
+func (v *VisitInfo) ShimSchema() shim.Schema                        { return v.shimSchema }
+func (v *VisitInfo) SchemaInfo() *SchemaInfo                        { return v.getSchemaInfo() }
+func (v *VisitInfo) EnclosingSchemaInfoMap() map[string]*SchemaInfo { return v.getSchemaInfoMap() }
+func (v *VisitInfo) EnclosingSchemaShimMap() shim.SchemaMap         { return v.getShimSchemaMap() }
+
+type VisitRoot interface {
+	PulumiToken() tokens.Type
+
+	isVisitRoot()
+}
+
+// This ensures that we can peform casts from VisitRoot to T for each valid root. Changing
+// this from T to *T is a breaking change.
+var (
+	_ VisitRoot = VisitResourceRoot{}
+	_ VisitRoot = VisitDataSourceRoot{}
+	_ VisitRoot = VisitConfigRoot{}
+)
+
+type VisitResourceRoot struct {
+	token tokens.Type
+
+	Info *ResourceInfo
+
+	TfToken string
+}
+
+func (r VisitResourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitResourceRoot) isVisitRoot()             {}
+
+type VisitDataSourceRoot struct {
+	token tokens.Type
+
+	Info *DataSourceInfo
+
+	TfToken string
+}
+
+func (r VisitDataSourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitDataSourceRoot) isVisitRoot()             {}
+
+type VisitConfigRoot struct{ token tokens.Type }
+
+func (r VisitConfigRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitConfigRoot) isVisitRoot()             {}
+
+type Visitor = func(VisitInfo) (VisitResult, error)
+
+// A wrapper around [TraverseSchemas] that panics on error. See [TraverseSchemas] for
+// documentation.
+func MustTraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor) {
+	err := TraverseSchemas(prov, traversalID, visitor)
+	contract.AssertNoErrorf(err, "failed to traverse provider schemas")
+}
+
+// _Efficiently_ traverse all schemas in a provider.
+//
+// TraverseSchemas operates in two stages:
+//
+//   - During `make tfgen` time, all fields in the provider are visited. TraverseSchemas
+//     builds a list of fields where `VisitResult{ HasEffect: true }` is returned.
+//
+//   - During the provider's normal runtime, only "effectful" fields are visited.
+//
+// traversalId must be unique for each traversal within a `make tfgen`, but the same
+// between `make tfgen` and the provider runtime. A simple descriptive string like `"apply
+// labels"` works great here.
+func TraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor, opts ...TraversalOption) error {
+	prov.MetadataInfo.assertValid() // We must have metadata info for an efficient (sparse) traversal.
+
+	var errs []error
+
+	var traverseOptions traversalOptions
+	for _, opt := range opts {
+		opt(&traverseOptions)
+	}
+
+	// forEffect indicates that the traversal is being run for side effects only, so
```

Comment:
The caller can provide an override for `forEffect`, but the default is to run in `forEffect` mode when `currentRuntimeStage == runningProviderStage`. See:

https://github.com/pulumi/pulumi-terraform-bridge/blob/a25a2fee3c454d02da84a1c46239da4a1812b40a/pkg/tfbridge/walk.go#L408-L412

>  Did you include it here because the caller is in the provider code?

I'm not sure what you mean here. Can you rephrase?

Reply:
That makes perfect sense - I was trying to confirm that the reason for the default is that provider code calls this function. Hence you can't set the default at the calling place.